src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -1073,4 +1073,67 @@ public void deleteTasksWithSolutionEntries(Long exerciseId) {
         programmingExerciseTaskRepository.deleteAll(tasks);
         programmingExerciseSolutionEntryRepository.deleteAll(solutionEntries);
     }
+
+    /**
+     * Locks or unlocks the repository if necessary due to the changes in the programming exercise
+     * @param programmingExerciseBeforeUpdate the original exercise with unchanged values
+     * @param updatedProgrammingExercise the updated exercise with new values
+     */
+    public void handleRepoAccessRightChanges(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
+        if (!programmingExerciseBeforeUpdate.isReleased()) {
+            if (updatedProgrammingExercise.isReleased() && Boolean.TRUE.equals(updatedProgrammingExercise.isAllowOfflineIde())) {
+                // There might be some repositories that have to be unlocked
+                unlockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+            return;
+        }
+        if (!updatedProgrammingExercise.isReleased()) {
+            if (Boolean.TRUE.equals(programmingExerciseBeforeUpdate.isAllowOfflineIde())) {
+                // Hide exercise again and lock repos
+                lockAllRepositories(programmingExerciseBeforeUpdate.getId());
+            }
+            return;
+        }
+
+        boolean lockedUnlockedRepos = handleRepoAccessRightChangesDueDates(programmingExerciseBeforeUpdate, updatedProgrammingExercise);
+        if (lockedUnlockedRepos) {
+            return;
+        }
+
+        handleRepoAccessRightChangesChangesOfflineIDE(programmingExerciseBeforeUpdate, updatedProgrammingExercise);
+    }
+
+    private boolean handleRepoAccessRightChangesDueDates(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
+        if (Boolean.TRUE.equals(updatedProgrammingExercise.isAllowOfflineIde())) {
+            ZonedDateTime now = ZonedDateTime.now();
+
+            if (programmingExerciseBeforeUpdate.getDueDate() != null && programmingExerciseBeforeUpdate.getDueDate().isBefore(now)
+                    && (updatedProgrammingExercise.getDueDate() == null || updatedProgrammingExercise.getDueDate().isAfter(now))) {
+                // New due date allows students to continue working on exercise
+                unlockAllRepositories(programmingExerciseBeforeUpdate.getId());
```

Comment:
I think individual due dates can also be earlier, so we also need to check them here.